torch_geometric/nn/models/gnn_llm.py

```diff
@@ -0,0 +1,449 @@
+        prepare_model_for_kbit_training,
+    )
+    WITH_PEFT = True
+except ImportError as e:  # noqa
+    WITH_PEFT = False
+
+try:
+    from transformers import AutoModelForCausalLM, AutoTokenizer
+    WITH_TRANSFORMERS = True
+except ImportError as e:  # noqa
+    WITH_TRANSFORMERS = False
+
+from torch_geometric.data import Batch
+from torch_geometric.nn.models import GAT
+from torch_geometric.utils import scatter
+
+BOS = '<s>[INST]'
+EOS_USER = '[/INST]'
+EOS = '[/s]'
+IGNORE_INDEX = -100
+llama2_str_name = "meta-llama/Llama-2-7b-chat-hf"
+gemma_str_name = "google/gemma-7b"
+max_txt_len = 512
+max_new_tokens = 32
+pad_token_id = 0
+padding_side = 'left'
+
+
+def get_llm_kwargs(mem_needed):
+    assert torch.cuda.is_available(), "GPU needed to run LLMs efficiently!"
+    avail_gpus = torch.cuda.device_count()
+    kwargs = {
+        "revision": "main",
+    }
+    max_mem_dict = {}
+    avail_mem_dict = {}
+    mem_total = 0
+    gpus_2_use_4_llm = 0
+    for i in range(avail_gpus):
+        available_mem = int(torch.cuda.mem_get_info(i)[0] // 1024**3)
+        mem_total += available_mem
+        avail_mem_dict[i] = available_mem
+        gpus_2_use_4_llm += 1
+        # We want to use the minimum number of GPUs that LLM can fit on
+        # this is to minimize the need for interGPU communications
+        if mem_total >= mem_needed:
+            break
+
+    for i in range(gpus_2_use_4_llm):
+        max_mem_dict[i] = str(avail_mem_dict[i]) + "GiB"
+    kwargs["max_memory"] = max_mem_dict
+    kwargs["device_map"] = "auto"
+    return kwargs
+
+
+class LLM(nn.Module):
+    def __init__(self, llm_name: str = "llama2", llm_dtype=torch.bfloat16,
+                 num_params: int = 7):
+        super().__init__()
+        if llm_name == "llama2":
+            self.printable_llm_name = "LLAMA2"
+            self.huggingface_str = llama2_str_name
+        elif llm_name == "gemma":
+            self.printable_llm_name = "GEMMA"
+            self.huggingface_str = gemma_str_name
+        else:
+            self.printable_llm_name = llm_name
+            self.huggingface_str = llm_name
+        self.mem_needed = 85 * num_params / 7
+        self.llm_dtype = llm_dtype
+        print('Loading ' + str(self.printable_llm_name))
+        kwargs = get_llm_kwargs(self.mem_needed)
+        print("Setting up " + self.printable_llm_name + " w/ kwargs =", kwargs)
+        self.tokenizer = AutoTokenizer.from_pretrained(self.huggingface_str,
+                                                       use_fast=False)
+        self.tokenizer.pad_token_id = pad_token_id
+        self.tokenizer.padding_side = padding_side
+        self.llm = AutoModelForCausalLM.from_pretrained(
+            self.huggingface_str, torch_dtype=self.llm_dtype,
+            low_cpu_mem_usage=True, **kwargs)
+        self.llm_device = self.llm.device
+        self.word_embedding = self.llm.model.get_input_embeddings()
+
+    def encode_inputs(self, samples: Batch):
```

Comment:
Can we drop the `Batch` dependency here and clearly define the expected function header?

Reply:
yes